openforcefield/utils/utils.py

```diff
@@ -274,6 +294,130 @@ def string_to_quantity(quantity_string):
+
+    Returns
+    -------
+    converted_smirnoff_data : dict
+        A hierarchical dict structured in compliance with the SMIRNOFF spec,
+        with quantity-defining strings converted to simtk.unit.Quantity objects
+    """
+    if isinstance(smirnoff_data, dict):
+        for key, value in smirnoff_data.items():
+            smirnoff_data[key] = convert_smirnoff_data_strings_to_quantity(value)
+        obj_to_return = smirnoff_data
+
+    elif isinstance(smirnoff_data, list):
+        for index, item in enumerate(smirnoff_data):
+            smirnoff_data[index] = convert_smirnoff_data_strings_to_quantity(item)
+        obj_to_return = smirnoff_data
+    else:
+        try:
+            obj_to_return = object_to_quantity(smirnoff_data)
+        except AttributeError:
+            obj_to_return = smirnoff_data
+        except TypeError:
+            obj_to_return = smirnoff_data
+        except SyntaxError:
+            obj_to_return = smirnoff_data
+
+    return obj_to_return
+
+
+def convert_smirnoff_data_quantity_to_string(smirnoff_data):
+    """
+    Traverses a SMIRNOFF data structure, attempting to convert all
+    quantities into strings.
+
+    Parameters
+    ----------
+    smirnoff_data : dict
+        A hierarchical dict structured in compliance with the SMIRNOFF spec
+
+    Returns
+    -------
+    converted_smirnoff_data : dict
+        A hierarchical dict structured in compliance with the SMIRNOFF spec,
+        with simtk.unit.Quantitys converted to string
+    """
+
+    if isinstance(smirnoff_data, dict):
+        for key, value in smirnoff_data.items():
+            smirnoff_data[key] = convert_smirnoff_data_quantity_to_string(value)
+        obj_to_return = smirnoff_data
+    elif isinstance(smirnoff_data, list):
+        for index, item in enumerate(smirnoff_data):
+            smirnoff_data[index] = convert_smirnoff_data_quantity_to_string(item)
+        obj_to_return = smirnoff_data
+    elif isinstance(smirnoff_data, unit.Quantity):
+        obj_to_return = quantity_to_string(smirnoff_data)
+    else:
+        obj_to_return = smirnoff_data
+
+    return obj_to_return
+
+
+def object_to_quantity(object):
+    """
+    Attempts to turn the provided object into simtk.unit.Quantity(s). Can handle strings, quantities, or iterators over
+    the same. Raises an exception if unable to convert all inputs
+
+    Parameters
+    ----------
+    object : string, quantity, or iterator of strings of quantities
+
+    Returns
+    -------
+    converted_object : simtk.unit.Quantity or list of simtk.unit.Quantity
+
+    """
+    from simtk import unit
+    if isinstance(object, unit.Quantity):
+        return object
+    elif isinstance(object, str):
+        return string_to_quantity(object)
+    else:
+        return [object_to_quantity(sub_obj) for sub_obj in object]
+
+def assert_object_units_are_compatible(object_name, object, unit_to_check, context=None):
```

Comment:
I'd expect an assertion error from this function. We could consider inheriting from `IncompatibleUnitError(AssertionError)` or change the name to `check_units_are_compatible()`.

Reply:
Huh, I had never thought about exactly what was raised by assert statements. Today I learned that `AssertionError` was a thing. I've changed this to `check_units_are_compatible()`.